### PR TITLE
Refactoring of plan building

### PR DESCRIPTION
**Pull request by systay**
This is a rebased #7033 

Today, in the planbuilding package, we have `builder`s and a `primitiveBuilder`. 

The primitive builder contains the current `builder` we are working on, the symtab,  jointab and a context object that allows us to talk with the vschema during planning.

Part of the logic lives in static methods, such as `buildSelectPlan` et al, that create the primitive builder and control the flow during planning. 
The other part of of the logic is spread out amongst the `builder`s - methods such as `PushFilter, PushSelect, MakeDistinct` and others.

During planning, the current plan is represented by a tree of `builder`s, that at the end of it all are turned into `engine.Primitive`s.

I think we can design this so it becomes easier to understand and work with:
- The `builder` interface is shock full of Liskov Substitution Principle problems - no implementor implements all interface methods. The structs implement the interface but most methods return an "not implemented" error.
- When adding new functionality or working with a bug, we often consider one aspect across all `builder`s - the bug is with DISTINCT, so we want to look at `MakeDistinct()`. Since we have 11 different `builder`s, this means we have to jump around to a lot of files, and it's harder to get an overview of how DISTINCT is solved.
- The `PushFilter`, `MakeDistinct`, `PushSelect` methods on the builder interface have very different signatures and behaviours - some change the builder in the primitiveBuilder, some return a new builder, and some are just mutating their own state.

This refactoring will try to address this by making the following changes:
- Extract most methods from the `builder` structs - for example, all `PushFilter` logic will instead live in `filtering.go` in a single method. The methods that will be extracted are:
   - PushFilter
   - PushSelect
   - MakeDistinct
   - PushGroupBy
   - PushOrderBy
   - SetUpperLimit
   - PushMisc
   - PushLock
- All extracted methods will take a primitiveBuilder and a builder as parameters (other params as needed by method), and return a builder.
- After removing these methods from the builder interface, we can rename it to `logicalPlan` or something along those lines.

### CODE REVIEW

File: go/vt/vtgate/planbuilder/filtering.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		return node, nil
+	case *vindexFunc:
+		return filterVindexFunc(node, filter)
+	case *mergeSort, *pulloutSubquery:
+		filteredInput, err := planFilter(pb, node.Inputs()[0], filter, whereType, origin)
+		if err != nil {
+			return nil, err
+		}
+		err = node.Rewrite(filteredInput)
+		return node, err
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/filtering.go:74]: please add a test to cover this case

File: go/vt/vtgate/planbuilder/grouping.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+		newInput, err := planGroupBy(pb, node.input, groupBy)
+		if err != nil {
+			return nil, err
+		}
+		node.input = newInput
+
+		return node, nil
+	case *join:
+		return nil, vterrors.New(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: group by on cross-shard join")
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/grouping.go:94]: please add a test to cover this case

File: go/vt/vtgate/planbuilder/grouping.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		inputs := node.Inputs()
+		input := inputs[0]
+
+		newInput, err := planDistinct(input)
+		if err != nil {
+			return nil, err
+		}
+		inputs[0] = newInput
+		node.Rewrite(inputs...)
+		return node, nil
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/grouping.go:112]: please add a test to cover this case

File: go/vt/vtgate/planbuilder/grouping.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		node.input = newInput
+		return node, nil
+
+	case *subquery:
+		return nil, vterrors.New(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: distinct on cross-shard subquery")
+	case *concatenate:
+		return newDistinct(node), nil
+	case *join:
+		return nil, vterrors.New(vtrpcpb.Code_UNIMPLEMENTED, "unsupported: distinct on cross-shard join")
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/grouping.go:139]: please add a test to cover these cases

**Review by harshit-gangal — APPROVED** [review]

**Review by sougou — COMMENTED** [review]
I think this is better than the old approach from the perspective of understanding how each SQL construct is handled. On the downside, it reduces the understanding of all the functionality that builds a primitive. A perfect code structure is not possible because we essentially have a matrix of primitives and AST constructs, and we have to provide code snippets for each of those.

However, if we have to make a choice, leaning towards grouping functions by the SQL construct (this new approach) feels better. So, I'm good with this.

Once review comments are addressed, you can push this forward.

**Pull request merged**